R/combine_checkboxes.R

```diff
@@ -139,17 +139,30 @@ combine_checkboxes <- function(supertbl,
 get_metadata_spec <- function(metadata_tbl,
                               selected_cols,
                               names_prefix,
-                              names_suffix,
-                              names_sep) {
+                              names_sep,
+                              names_glue) {
+  check_metadata_fields_exist(metadata_tbl, selected_cols)
+
   # Create a metadata reference table linking field name to raw and label values
-  out <- metadata_tbl %>%
-    filter(.data$field_name %in% selected_cols) %>%
-    mutate(
-      .value = sub("___.*$", "", .data$field_name),
-      .new_value = case_when(!is.null(names_suffix) ~ paste(names_prefix, .value, names_suffix, sep = names_sep),
-        .default = paste(names_prefix, .data$.value, sep = names_sep)
+  if (!is.null(names_glue)) {
+    check_installed("glue", reason = "to use `names_glue` in `combine_checkboxes()`")
+    # Similar to pivot_*, use of `names_glue` overrides use of names_prefix/sep
+    out <- metadata_tbl %>%
+      filter(.data$field_name %in% selected_cols) %>%
+      mutate(
+        .value = sub("___.*$", "", .data$field_name),
+        .new_value = as.character(glue::glue(names_glue))
```

Comment:
I think we need to be more careful with this implementation of `names_glue`.

Right now we're giving `names_glue` access to everything in the metadata which can cause weird results:

```r
db_label |>
  combine_checkboxes("demographics", starts_with("race"), names_glue = "xyz_{field_name}") |>
  extract_tibble("demographics")
```

In this example we actually create the wrong number of output columns silently.

I think we should:
1. Use `glue_data()` rather than `glue()` to scope what the user actually has access to glue with:

```r
# Could include more things we want to give the user access to here
glue_env <- select(out, .value)

out <- out |>
  mutate(.new_value = glue_data(glue_env, names_glue))
```

2. Enforce the constraint that `.new_value` is the same within each level of `.value`. This ensures the user always gets the expected number of output columns.

Reply:
@ezraporter quick q. Happy to use `glue_data()` as discussed, but in this example the output is still going to be the same since `field_name` is still accessible from `out`. The intention here is to have it fail instead of falsely grabbing the `field_name` from the metadata (for now) right? See if this is the change we're looking for for (1):

```r
get_metadata_spec <- function(metadata_tbl,
                              selected_cols,
                              names_prefix,
                              names_sep,
                              names_glue) {
  check_metadata_fields_exist(metadata_tbl, selected_cols)

  # Create a metadata reference table linking field name to raw and label values
  out <- metadata_tbl %>%
    filter(.data$field_name %in% selected_cols) %>%
    mutate(
      .value = sub("___.*$", "", .data$field_name)
    )

  if (!is.null(names_glue)) {
    # Similar to pivot_*, use of `names_glue` overrides use of names_prefix/sep
    glue_env <- select(out, .value) %>%
      mutate(.new_value = as.character(glue::glue_data(., names_glue))) %>%
      select(.new_value)

    out <- cbind(out, glue_env)
  } else {
    out <- out %>%
      mutate(
        .new_value = case_when(names_prefix != "" ~ paste(names_prefix, .value, sep = names_sep),
          .default = paste(names_prefix, .data$.value, sep = "")
        )
      )
  }
```

Reply:
> 2. Enforce the constraint that .new_value is the same within each level of .value. This ensures the user always gets the expected number of output columns.

See what you think of the most recent commit for this (1).

Reply:
I'm not sure I understand why we're converting things to factors. The validation I was talking about was something like:

```r
check <- out |>
  group_by(.value) |>
  summarize(n=n_distinct(.new_value)) |>
  pull(n)

if (!all(check == 1)) {
  # Throw an error
}
```

Reply:
Ah ok, I thought you wanted more of an enforcement not a check. Can implement. 

Out of curiosity what's an example with this set up of how someone would still trigger this?

Reply:
Check function implemented, but unsure what the message should be so let me know how you'd like to tweak.

```r
  error_data <- tibble::tribble(
    ~"id", ~"col1", ~"col2",
    1, "A", "A1",
    2, "B", "B1",
    3, "B", "B2"
  )

  check_equal_col_summaries(error_data, col1, col2)
```

> Error in `check_equal_col_summaries()`:
✖ Encountered unequal naming outputs.
! `combine_checkboxes()` call resulted in column output: `A`, `B`, and `B` and new column output: `A1`, `B1`, and `B2`.
Run `rlang::last_trace()` to see where the error occurred.

Reply:
> Out of curiosity what's an example with this set up of how someone would still trigger this?

Mmm I see. I think in our current set-up it can't get triggered so maybe it's redundant at this point? There are some weird cases like this but it's pretty contrived:

```r
data <- tibble::tibble(.value = c("A", "A", "B"))
vector_in_env <- 1:3

data |>
  mutate(.new_value = glue_data(data, "{vector_in_env}_{.value}")
```

All that is to say: maybe we should just drop it. If we keep it we probably want it to say something like "Checkbox field `B` resulted in multiple output columns, `B1` and `B2`. Check that `names_glue` defines only 1 output column for each checkbox field."

Reply:
Ok, yea I couldn't think of a way from the UI to trigger this but there's no real harm in keeping it in for now. I can put a comment in that says we may forgo it in the future. I'll update the error message to be closer to your suggestion.